Introduce gsheets connection into st-connection-prpr

sfc-gh-tszerszen opened the pull request:
# Streamlit GSheets Connection

A prototype GSheets Connection for st.connection in Streamlit. 
GSheets Connection works in two modes:
* in Read Only mode, using publicly shared Spreadsheet URLs (Read Only mode)
* CRUD operations support mode, with Authentication using Service Account. In order to use Service Account mode you need to enable Google Drive and Google Sheets API in [Google Developers Console](https://console.developers.google.com/).
Follow **Initial setup for CRUD mode** section in order to authenticate your Streamlit app first.

## Initial setup for CRUD mode

1. Setup `.streamlit/secrets.toml` inside your Streamlit app root directory,  
check out [Secret management documentation](https://docs.streamlit.io/streamlit-community-cloud/get-started/deploy-an-app/connect-to-data-sources/secrets-management) for references.
2. [Enable API Access for a Project](https://docs.gspread.org/en/v5.7.1/oauth2.html#enable-api-access-for-a-project)
    * Head to [Google Developers Console](https://console.developers.google.com/) and create a new project (or select the one you already have).
    * In the box labeled “Search for APIs and Services”, search for “Google Drive API” and enable it.
    * In the box labeled “Search for APIs and Services”, search for “Google Sheets API” and enable it.
3. [Using Service Account](https://docs.gspread.org/en/v5.7.1/oauth2.html#for-bots-using-service-account)
    * Enable API Access for a Project if you haven’t done it yet.
    * Go to “APIs & Services > Credentials” and choose “Create credentials > Service account key”.
    * Fill out the form
    * Click “Create” and “Done”.
    * Press “Manage service accounts” above Service Accounts.
    * Press on ⋮ near recently created service account and select “Manage keys” and then click on “ADD KEY > Create new key”.
    * Select JSON key type and press “Create”.

You will automatically download a JSON file with credentials. It may look like this:
```
{
    "type": "service_account",
    "project_id": "api-project-XXX",
    "private_key_id": "2cd … ba4",
    "private_key": "-----BEGIN PRIVATE KEY-----\nNrDyLw … jINQh/9\n-----END PRIVATE KEY-----\n",
    "client_email": "473000000000-yoursisdifferent@developer.gserviceaccount.com",
    "client_id": "473 … hd.apps.googleusercontent.com",
    ...
}
```
Remember the path to the downloaded credentials file. Also, in the next step you’ll need the value of client_email from this file.
* **Very important!** Go to your spreadsheet and share it with a client_email from the step above. Just like you do with any other Google account. If you don’t do this, you’ll get a `gspread.exceptions.SpreadsheetNotFound` exception when trying to access this spreadsheet from your application or a script.

4. Inside `streamlit/secrets.toml` place `service_account` configuration from downloaded JSON file, in the following format (where `gsheets` is your `st.connection` name):
```
# .streamlit/secrets.toml

[connections.gsheets]
spreadsheet = "<spreadsheet-name-or-url>"
worksheet = "<worksheet-gid-or-folder-id>"  # worksheet GID is used when using Public Spreadsheet URL, when usign service_account it will be picked as folder_id
type = ""  # leave empty when using Public Spreadsheet URL, when using service_account -> type = "service_account"
project_id = ""
private_key_id = ""
private_key = ""
client_email = ""
client_id = ""
auth_uri = ""
token_uri = ""
auth_provider_x509_cert_url = ""
client_x509_cert_url = ""
```

## Publicly Shared Spreadsheet Example
```python
# example/st_app.py

import streamlit as st
from streamlit_gsheets import GSheetsConnection

url = "https://docs.google.com/spreadsheets/d/1JDy9md2VZPz4JbYtRPJLs81_3jUK47nx6GYQjgU8qNY/edit?usp=sharing"

conn = st.experimental_connection("gsheets", type=GSheetsConnection)

data = conn.read(spreadsheet=url, usecols=[0, 1])
st.dataframe(data)
```

## Service Account Example

```python
# example/st_app_gsheets_using_service_account.py

import streamlit as st
from streamlit_gsheets import GSheetsConnection

st.title("Read Google Sheet as DataFrame")

conn = st.experimental_connection("gsheets", type=GSheetsConnection)
df = conn.read(worksheet="Example 1")

st.dataframe(df)
```

```toml
# .streamlit/secrets.toml

[connections.gsheets]
spreadsheet = "<spreadsheet-name-or-url>"
worksheet = "<worksheet-gid-or-folder-id>"  # worksheet GID is used when using Public Spreadsheet URL, when usign service_account it will be picked as folder_id
type = ""  # leave empty when using Public Spreadsheet URL, when using service_account -> type = "service_account"
project_id = ""
private_key_id = ""
private_key = ""
client_email = ""
client_id = ""
auth_uri = ""
token_uri = ""
auth_provider_x509_cert_url = ""
client_x509_cert_url = ""
```

```txt
# requirements.txt

streamlit==1.22
git+https://github.com/sfc-gh-jcarroll/st-connection-prpr.git#subdirectory=gsheets_connection
pandasql  # this is for example/st_app.py only
```

**Note on install:** Streamlit 1.22 is not yet released; you can find a compatible .whl file
[here](https://core-previews.s3-us-west-2.amazonaws.com/pr-6457/streamlit-1.21.0-py2.py3-none-any.whl).

## Full example
Check gsheets_connection/example directory for full example of the usage.

## Q&A
* > Does this work with a public spreadsheet without the authentication details? Or only a private spreadsheet?

   GSheets Connection works in two modes:
  * in Read Only mode, using publicly shared Spreadsheet URLs (Read Only mode)
  * CRUD operations support mode, with Authentication using Service Account. In order to use Service Account mode you need to enable Google Drive and Google Sheets API in [Google Developers Console](https://console.developers.google.com/).
  Follow **Initial setup for CRUD mode** section in order to authenticate your Streamlit app first.

sfc-gh-zblackwood commented:
I'm curious about why you decided to use pandasql, which hasn't gotten any meaningful updates since 2016. Seems pretty risky to rely on such an old library. Did you consider https://github.com/betodealmeida/shillelagh?

sfc-gh-tszerszen commented:
> I'm curious about why you decided to use pandasql, which hasn't gotten any meaningful updates since 2016. Seems pretty risky to rely on such an old library. Did you consider https://github.com/betodealmeida/shillelagh?

It's only for the demo purposes, it was super easy to do the demo using `psql.load_births()` and `psql.load_meat()`, sorry if that is confusing. When it comes to query engine for Google Sheets I'm using DuckDB.

sfc-gh-tszerszen commented:
**Latest update:**
* I've just introduced optional `spreadsheet` and `spreadsheet_folder_id` variables in `.streamlit/secrets.toml`
* ...these optional variables allow to hardcode used Spreadsheet, so user does not have to specify it.
* Hence I reorganised order of parameters and made `spreadsheet` and `folder_id` optional.

sfc-gh-jcarroll commented on gsheets_connection/example/.streamlit/example.secrets.toml:
```diff
@@ -0,0 +1,15 @@
+# .streamlit/secrets.toml
+
+[connections.gspread]
```
In st_app.py the connection is named `gsheets` so I expect it wouldn't find this section `gspread`. Am I missing something?

sfc-gh-tszerszen replied:
I've just updated it 👍 it is now `gsheets` it was a typo on my side. I thought it through and decided to go with `gsheets` name since I've started to plan to extend standard `gspread` functionality to at least SQL support 👍

sfc-gh-jcarroll commented on gsheets_connection/streamlit_gsheets/gsheets_connection.py:
```diff
@@ -0,0 +1,1072 @@
+            **options,
+        )
+
+    def sql(
```
I would recommend to just support `query()` and not `sql()` unless there's some difference between the behavior.

sfc-gh-jcarroll commented on gsheets_connection/streamlit_gsheets/gsheets_connection.py:
```diff
@@ -0,0 +1,1072 @@
+            spreadsheet = self._get_default_spreadsheet()
+        if not folder_id and not (type(folder_id) is int and folder_id == 0):
+            folder_id = self._get_default_spreadsheet_folder_id()
+        return self.sql(
```
This should follow the pattern to be wrapped with `@cache_data` by default and support TTL argument in the method call

sfc-gh-tszerszen replied:
Will do 👍

sfc-gh-tszerszen replied:
It's done, I got rid of many methods (decided to get rid of the methods which do not operate on DataFrames), and two read methods `query` and `read` support `TTL` and `max_entries` args and cache their data 👍

sfc-gh-jcarroll commented on gsheets_connection/streamlit_gsheets/gsheets_connection.py:
```diff
@@ -0,0 +1,1072 @@
+            spreadsheet = self._get_default_spreadsheet()
+        if not folder_id and not (type(folder_id) is int and folder_id == 0):
+            folder_id = self._get_default_spreadsheet_folder_id()
+        return self.get_all_records(
```
This should follow the pattern to be wrapped with `@cache_data` by default and support TTL argument in the method call

sfc-gh-tszerszen replied:
It's done, I got rid of many methods (decided to get rid of the methods which do not operate on DataFrames), and two read methods `query` and `read` support `TTL` and `max_entries` args and cache their data 👍

sfc-gh-jcarroll commented on gsheets_connection/streamlit_gsheets/gsheets_connection.py:
```diff
@@ -0,0 +1,1072 @@
+)
+
+
+class GSheetsConnection(ExperimentalBaseConnection[Client]):
```
This is cool!!

Using duckdb to support query() is a really nice addition too.

I think the st.connection-y thing to do would be to just have `read()`, `query()`, and `client`. It seems like the other methods mostly just wrap functionality in client.

If it was me I would be concerned about testing / maintaining so many additional methods and how much use / value they will provide for developers vs just using the underlying client. But if you see value and feel comfortable supporting them going forward, I'm OK with it too.

sfc-gh-tszerszen replied:
Thank you Joshua your feedback is helpful and appreciated! 🥇 
I've just implemented the changes, so feel free to review it any time 🙂 

> I think the st.connection-y thing to do would be to just have `read()`, `query()`, and `client`. It seems like the other methods mostly just wrap functionality in client.

After thoughtful consideration I've decided to keep the methods which resemble CRUD operations and operate on DataFrame (I think they are valuable and helpful and it's worth supporting them).
Got rid of most other methods which do not operate on `DataFrame`, left `open_spreadsheet` and `select_worksheet`, since their are needed as functions inside other methods and will be useful to users for convenient access to the `gspread` client.

sfc-gh-tszerszen commented on gsheets_connection/streamlit_gsheets/gsheets_connection.py:
```diff
@@ -0,0 +1,1072 @@
+            spreadsheet = self._get_default_spreadsheet()
+        if not folder_id and not (type(folder_id) is int and folder_id == 0):
+            folder_id = self._get_default_spreadsheet_folder_id()
+        return self.sql(
```
It's done, I got rid of many methods (decided to get rid of the methods which do not operate on DataFrames), and two read methods `query` and `read` support `TTL` and `max_entries` args and cache their data 👍

sfc-gh-zblackwood commented on gsheets_connection/streamlit_gsheets/gsheets_connection.py:
```diff
@@ -0,0 +1,566 @@
+
+        Parameters
+        ------------
+        worksheet: str | int | Worksheet:
```
Would be good to add the default values to `worksheet`, `spreadsheet`, `ttl`

sfc-gh-tszerszen replied:
Thank you, I've just added `Defaults to "..."` on `worksheet`, `spreadsheet` and `ttl` in their docstrings 👍

sfc-gh-zblackwood commented on gsheets_connection/streamlit_gsheets/gsheets_connection.py:
```diff
@@ -0,0 +1,566 @@
+        **options,
+    ) -> DataFrame:
+        """Run SQL query against spreadsheet. Worksheet name should be used as
+        Table name inside double quota.
```
```suggestion
        Table name inside double quotes.
```

sfc-gh-tszerszen replied:
😅 Yeah, you're totally right, thank you, just fixed it!

sfc-gh-zblackwood commented on pages/02_🙌_Community_connections.py:
```diff
@@ -14,12 +14,18 @@
-st.dataframe(pet_owners)
+st.title("Read Google Sheet as DataFrame")
+
+spreadsheet = "Example Google Sheet created by st.connection example"  # make sure such spreadsheet exists
```
This confused me the first few times I looked at it -- could you change it to be `spreadsheet_name = `? I think that would make it more clear that this string is an actual name. I'm not sure why, but every time I look at that string, I just assume it's just a descriptive string that will be shown with `st.write`

sfc-gh-tszerszen replied:
It's renamed 👍

sfc-gh-zblackwood commented on gsheets_connection/streamlit_gsheets/utils.py:
```diff
@@ -0,0 +1,99 @@
+"""Code take from https://github.com/python-validators/validators/blob/master/validators/utils.py"""
```
Why not just include python-validators as a dependency? Seems reasonably light-weight, and seems like a reasonable option since you're copying two different files from it.

sfc-gh-tszerszen replied:
Good point, I've removed these files and put validators in `setup.py` as a dependency instead 👍

sfc-gh-tszerszen commented:
Thank you @sfc-gh-zblackwood for your latest review, I've commited changes suggested by you 🙂

sfc-gh-jcarroll commented on gsheets_connection/README.md:
````diff
@@ -0,0 +1,103 @@
+```toml
+# .streamlit/secrets.toml
+
+[connections.gspread]
````
Nit: change to `gsheets`

sfc-gh-tszerszen replied:
Sorry I've missed it, you mentioned it before! 🙇  It's fixed now 🙂

sfc-gh-jcarroll commented:
Nice this is great. Appreciate the updates.

Last question from me: Does this work with a public spreadsheet without the authentication details? Or only a private spreadsheet?

- if it only works with the authentication - call that out explicitly - and we should merge and talk later about how to support public spreadsheet easily
- If there's a way to support public spreadsheet - document it and add an example in the readme

sfc-gh-jcarroll commented:
BTW I want to look tomorrow at getting dedicated repos in streamlit org for these - if you want to move it towards a dedicated repo feel free to.

sfc-gh-tszerszen commented:
> Last question from me: Does this work with a public spreadsheet without the authentication details? Or only a private spreadsheet?
> 
> * if it only works with the authentication - call that out explicitly - and we should merge and talk later about how to support public spreadsheet easily
> * If there's a way to support public spreadsheet - document it and add an example in the readme

Updated README.md with the following comment:

GSheets Connection works only with Authentication using Service Account.
In order to use it you need to enable Google Drive and Google Sheets API in [Google Developers Console](https://console.developers.google.com/).
Follow Initial setup section in order to authenticate your Streamlit app first.

Also introduced Q&A section and allowed myself to copy your question there.

🤔 I did a quick research and do not see an easy method for working with public spreadsheets, feel free to share if you find one. In my spare time I'll look for solutions myself.

I'm happy to support this project and thank you for your reviews, appreciate your work! 🙂 (This is a great idea for `st.connection` extension and I'm sure it will be useful, I'm glad I can work on it).

sfc-gh-tszerszen commented:
@sfc-gh-zblackwood @sfc-gh-jcarroll thank you for your review and awesome input ❤️ Just added support for publicly shared Spreadsheet, it's an API change so please re-review it if you can 🙂 

I'll move it to repo called `gsheets_connection` inside Streamlit organisation after final look from you (if you can provide it 🙇 )

sfc-gh-zblackwood commented on gsheets_connection/streamlit_gsheets/gsheets_connection.py:
```diff
@@ -0,0 +1,822 @@
+
+    def read(
+        self,
+        worksheet_or_public_spreadsheet_url: Union[int, str, Worksheet] = "Sheet1",
```
It seems like having a public url go in the `spreadsheet` argument would be more natural. Is there a reason why that won't work well?

sfc-gh-tszerszen replied:
Will work well, I decided to do it that way because I'd like to be this first **positional** argument:
* when you're using `Service Account` your first argument is `Worksheet` name, so you do not have to provide a keyword
* when you're using Read Only Mode your first argument is an URL

IDK maybe it'd more intuitive to move it to Spreadsheet argument, or make Spreadsheet argument support URLs. What do you think? 🤔 

I don't have strong opinions here I think it's tricky

sfc-gh-tszerszen replied:
@sfc-gh-zblackwood for now I've just updated the code in such a way, that spreadsheet can also be an URL to publicly shared Worksheet, so one can provide it as both parameters

sfc-gh-jcarroll replied:
Agree with Zachary very often you use a spreadsheet with no worksheet

sfc-gh-jcarroll replied:
I think I would do no positional args, only keyword args. so `conn.read()` will work if you put the spreadsheet in secrets, or `conn.read(spreadsheet="https://...")` will also work.

This is also nicely aligned to the files API where you do `conn.read(csv="s3://...")`

sfc-gh-tszerszen replied:
@sfc-gh-jcarroll thank you for your input, I've just updated GSheets Connection and all arguments are keyword only and work as you suggested 👍

sfc-gh-zblackwood commented on gsheets_connection/streamlit_gsheets/gsheets_connection.py:
```diff
@@ -0,0 +1,822 @@
+        """
+        if not self.is_service_account:
+            msg = "Method `update` is not supported when not using Service Account authentication."
+            raise UnsupportedOperationException(msg)
```
Why is this necessary? Couldn't you simply allow Google Sheets to raise whatever errors it does (and possible catch them if you'd like). But, if someone enables editing by url, why not allow them to use this connection to also edit it, like we show in the current tutorial https://docs.streamlit.io/knowledge-base/tutorials/databases/public-gsheet?

sfc-gh-tszerszen replied:
I wanted errors to be descriptive, easy, short and intuitive however maybe it is throwing out the baby with the bathwater 🤔 I can just remove these lines if you prefer

sfc-gh-tszerszen replied:
I think we can also consider using here `StreamlitAPIException` so StackTraces are hidden

sfc-gh-zblackwood replied:
I don't have any issue with raising custom errors, I just would prefer to do something like this:

```python
try:
    try to do updates
except APIError as e:
    if "write failed" in str(e): # or whatever the gspread one returns if you try to write
        raise UnsupportedOperationException("This spreadsheet cannot be written to without using a service account")
   raise e
```

sfc-gh-tszerszen replied:
@sfc-gh-zblackwood got it, I think it'd better to just leave exceptions as they are and I've just updated the code and removed these custom exceptions on CRUD operations.

Please let me know if you're ok with it

sfc-gh-jcarroll commented on gsheets_connection/README.md:
````diff
@@ -0,0 +1,129 @@
+```python
+# example/st_app.py
+
+import streamlit as st
+from streamlit_gsheets import GSheetsConnection
+
+url = "https://docs.google.com/spreadsheets/d/1JDy9md2VZPz4JbYtRPJLs81_3jUK47nx6GYQjgU8qNY/edit?usp=sharing"
+
+conn = st.experimental_connection("gsheets", type=GSheetsConnection)
+
+data = conn.read(url)
+st.dataframe(data)
````
I love this!!!!!

sfc-gh-zblackwood commented on gsheets_connection/README.md:
```diff
@@ -0,0 +1,129 @@
+
+conn = st.experimental_connection("gsheets", type=GSheetsConnection)
+
+data = conn.read(spreadsheet=url, usecols=[0, 1])
```
nit: use either df or data consistently in these two examples. I'd go with df as a small hint that it's specifically a dataframe being returned.

sfc-gh-tszerszen replied:
Thank you, just updated it 👍

sfc-gh-zblackwood commented:
Weird thing -- this doesn't work for some reason -- it's a publicly shared sheet 
```
url = "https://docs.google.com/spreadsheets/d/14vDbxWk35-p8ykQvhQgmxkcqvKj6twro3nq_xsF6-TE/edit?usp=sharing"
```
Any idea why, @sfc-gh-tszerszen ?

sfc-gh-tszerszen commented:
> https://docs.google.com/spreadsheets/d/14vDbxWk35-p8ykQvhQgmxkcqvKj6twro3nq_xsF6-TE/edit?usp=sharing

Yes, I was too strict when validating URLs and extracting Spreadsheet IDs Big Kudos for finding it out! 🙇 It should be fixed now, can you please check it out again?

![image](https://user-images.githubusercontent.com/78742618/234712048-29db6927-e547-40d1-94ef-fdc72af5b9e2.png)

sfc-gh-jcarroll commented on gsheets_connection/streamlit_gsheets/gsheets_connection.py:
```diff
@@ -0,0 +1,729 @@
+    def query(
+        self,
+        *,  # keyword-only arguments:
+        sql: str,
```
Ah - I think `sql` could be a positional arg? is there a situation where you wouldn't specify it? I can't think of one

Otherwise this looks good to me!

sfc-gh-tszerszen replied:
No problem, just updated the PR and now `sql` is first positional argument of `query` method 🙂

sfc-gh-zblackwood replied:
@sfc-gh-tszerszen Excellent! Reading now works fine. The one thing that I would love to see still would be to be able to write to a public spreadsheet without needing to use credentials. For example, I have given anybody with this link https://docs.google.com/spreadsheets/d/14vDbxWk35-p8ykQvhQgmxkcqvKj6twro3nq_xsF6-TE/edit#gid=0 permission to write to the sheet. That provides a super simple database that is easy to set up, and is secure-by-obscurity, which is sufficient for many needs. To be clear, I'm not sure if this is possible, or feasible, using gspread, but if it is doable I would love to see that addition.

sfc-gh-tszerszen replied:
> ...permission to write to the sheet. That provides a super simple database that is easy to set up, and is secure-by-obscurity, which is sufficient for many needs.

Totally agree and I'd love to see that too ❤️ , I'd like to copy/paste our internal discussion here for future reference:

It is technically possible, because Google allows the browser to do it using the link, however:
* Libraries and Google API do not officially support Reading (I extracted URL for CSV download myself, not to mention update/delete)
* so we can implement this, but if Google changes anything in the API we need to reinvent the wheel again
* and we need to find out manually API calls

**So technically possible, but requires maintenance, is not supported by Google officially and requires reverse engineering.**
On the other hand Service Account is officially supported and API available.

vdonato commented on gsheets_connection/README.md:
```diff
@@ -0,0 +1,129 @@
+# Streamlit GSheets Connection
+
+A prototype GSheets Connection for st.connection in Streamlit. 
```
nit: `st.connection` -> `st.experimental_connection`

vdonato commented on gsheets_connection/README.md:
```diff
@@ -0,0 +1,129 @@
+check out [Secret management documentation](https://docs.streamlit.io/streamlit-community-cloud/get-started/deploy-an-app/connect-to-data-sources/secrets-management) for references.
+2. [Enable API Access for a Project](https://docs.gspread.org/en/v5.7.1/oauth2.html#enable-api-access-for-a-project)
+    * Head to [Google Developers Console](https://console.developers.google.com/) and create a new project (or select the one you already have).
+    * In the box labeled “Search for APIs and Services”, search for “Google Drive API” and enable it.
+    * In the box labeled “Search for APIs and Services”, search for “Google Sheets API” and enable it.
```
It might be useful to combine these two points into one line (or maybe have the APIs to enable be nested bullet points). I initially thought it was one duplicated line because they're so similar

sfc-gh-tszerszen replied:
Combined into one 👍

vdonato commented on gsheets_connection/streamlit_gsheets/gsheets_connection.py:
```diff
@@ -0,0 +1,746 @@
+        """
+        return self.client.set_default(spreadsheet=spreadsheet, worksheet=worksheet)
+
+    def _repr_html_(self) -> str:
+        module_name = getattr(self, "__module__", None)
+        class_name = type(self).__name__
+        if self._connection_name:
+            name = f"`{self._connection_name}` "
+            if len(self._secrets) > 0:
+                cfg = f"""- Configured from `[connections.{self._connection_name}]`
+                """
+            else:
+                cfg = ""
+        else:
+            name = ""
+            cfg = ""
+        md = f"""
+        ---
+        **st.connection {name}built from `{module_name}.{class_name}`**
+        {cfg}
+        - Learn more using `st.help()`
+        ---
+        """
+        return md
```
I think this can be removed since the default implementation of `ExperimentalBaseConnection` now does more or less the same thing

sfc-gh-tszerszen replied:
Removed 👍

vdonato reviewed (commented):
Left a few minor comments

I'll leave the proper code review to @sfc-gh-zblackwood and @sfc-gh-jcarroll since I'm not super familiar with the `gspread` library, but overall everything looked reasonable to me in my quick read-through

sfc-gh-tszerszen commented:
@vdonato thank you for your review I've applied the changes 👍 

**Everyone**, gsheets-connection main repository has moved to: https://github.com/streamlit/gsheets-connection, please provide reviews there instead
